### PR TITLE
initial steps for `%in%`

### PR DESCRIPTION
**Pull request by romainfrancois**
This needs more work to match #88 but it's a start. The trick is that the rhs is wrapped into a constant list: 

``` r
duckdbrfuns:::rfuns("%in%", data.frame(a = c(1, 2, 4)), list(c(1, 2, 3)))
#> [1]  TRUE  TRUE FALSE
```

``` r
duckdbrfuns:::rfuns("%in%", data.frame(a = c(1L, 3L)), list(c(1L, 2L)))
#> [1]  TRUE FALSE
```

``` r
duckdbrfuns:::rfuns("%in%", data.frame(a = c(TRUE, FALSE)),  list(c(TRUE)))
#> [1]  TRUE FALSE
```

<sup>Created on 2024-05-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: src/in.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+void in(DataChunk &args, ExpressionState &state, Vector &result) {
+	auto y = args.data[1];
+	auto size = ListVector::GetListSize(y);
+	auto y_data = FlatVector::GetData<T>(ListVector::GetEntry(y));
+
+	// TODO: handle NA in set
+	auto set = unordered_set<T>();
+	for (int i = 0; i < size; i++) {
+		set.insert(y_data[i]);
+	}
```
Comment (romainfrancois) [inline src/in.cpp:24]: This uses `std::unordered_set<>` but `duckdb` probably has its own set/map classes ?

Comment (krlmlr) [reply src/in.cpp:24]: Perhaps review how `list_contains()` is implemented in DuckDB, and adapt to use R semantics?

https://duckdb.org/docs/sql/functions/nested.html#list_containslist-element

Comment (romainfrancois) [reply src/in.cpp:24]: I followed the 🐇 to https://github.com/duckdb/duckdb-r/blob/3ede6f11178e2e6c6ae0e531735ab1ead650cc61/src/duckdb/src/include/duckdb/function/scalar/list/contains_or_position.hpp#L6

From what I understand though `list_contains()` only handles the case where the rhs is one value, i.e. : 

``` r
library(duckdbrfuns)
library(tibble)

con <- duckdbrfuns:::duckdb_con()
duckdb <- asNamespace("duckdb")

experimental <- FALSE
df1 <- tibble(a = list(1:10))

rel1 <- duckdb$rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb$rel_project(
  rel1,
  list(
    duckdb$expr_reference("a"),
    duckdb$expr_function("list_contains", list(
      duckdb$expr_reference("a"),
      duckdb$expr_constant(42)
    ))
  
  )
)
rel2
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a, list_contains(a, 42.0) as list_contains(a, 42.0)]
#>   r_dataframe_scan(0x1363c7de0)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (INTEGER[])
#> - list_contains(a, 42.0) (BOOLEAN)
```

<sup>Created on 2024-05-27 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

This is good inspiration, I'll get back to it later this week. Thanks for the hint @krlmlr .

Comment (romainfrancois) [reply src/in.cpp:24]: Actually it is the other way around:

``` r
library(duckdbrfuns)
library(tibble)

con <- duckdbrfuns:::duckdb_con()
duckdb <- asNamespace("duckdb")

experimental <- FALSE
df1 <- tibble(a = 1:10)
values <- c(1, 2, 3)


rel1 <- duckdb$rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb$rel_project(
  rel1,
  list(
    duckdb$expr_reference("a"),
    duckdb$expr_function("list_contains", list(
      duckdb$expr_constant(list(values)),
      duckdb$expr_reference("a")
    ))
  )
)
duckdb$rel_to_altrep(rel2)[, 2][]
#>  [1]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
```

So if we were happy about only comparing columns to list of the same types, we could just edit the `"%in%"` translation in `duckplyr`, but that does not work across types: 

``` r
library(duckdbrfuns)
library(tibble)

con <- duckdbrfuns:::duckdb_con()
duckdb <- asNamespace("duckdb")

experimental <- FALSE
df1 <- tibble(a = 1:10)
values <- c("1", "2")


rel1 <- duckdb$rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb$rel_project(
  rel1,
  list(
    duckdb$expr_reference("a"),
    duckdb$expr_function("list_contains", list(
      duckdb$expr_constant(list(values)),
      duckdb$expr_reference("a")
    ))
  )
)
#> Error: {"exception_type":"Binder","exception_message":"Cannot get list_position of element of type INTEGER in a list of type VARCHAR[] - an explicit cast is required"}
```

so a solution that indeed takes inspiration on the implementation of `list_contains` but adapts it to some of the work we did in relop that compares across types could work.

**Comment by romainfrancois** [comment]
Updated based on what `list_index` does. Now handles different types, i.e. `<chr> %in% <dbl>` using the tools from relop

``` r
duckdbrfuns:::rfuns("%in%", data.frame(a = c("1", "2", "4")), list(c(1, 2, 3)))
#> [1]  TRUE  TRUE FALSE
```

This needs some more care to handle `NA` properly, i.e. the same way as R: 

```r
> c(1, NA, 3) %in% c(1, 2, 3)
[1]  TRUE FALSE  TRUE
> duckdbrfuns:::rfuns("%in%", data.frame(a = c(1, NA, 3)), list(c(1, 2, 3)))
[1] TRUE   NA TRUE
```

so we probably need to not use `UnaryExecutor::ExecuteWithNulls` but use a raw loop instead like we did for `is.na()`.

**Review by krlmlr — COMMENTED** [review]
Thanks. Testing corner cases here seems even more important than for other functions, in particular with `NA`. I've spent a lot of time chasing those with revdepchecks. They are now materialized as tests in duckplyr, so we'll notice if anything is missing. Still, better to test at the source.

https://github.com/duckdblabs/duckplyr/blob/7bc1ccb925ef742cf7805c9391fec19b1233123c/tools/00-funs.R#L618-L625

**Comment by romainfrancois** [comment]
I'm going to let this sleep for a while, until we simplify the relop aka spaceship to only handle same type comparison, or at least not do the string vs number and then I'll come back here.

File: src/relop.cpp [hunk 2]
```diff
@@ -264,5 +264,241 @@ ScalarFunctionSet base_r_gte() {
...
+	// set.AddFunction(IN_VARIANT(VARCHAR, INTEGER));
+	// set.AddFunction(IN_VARIANT(INTEGER, VARCHAR));
+	// set.AddFunction(IN_VARIANT(VARCHAR, DOUBLE));
+	// set.AddFunction(IN_VARIANT(DOUBLE, VARCHAR));
+	set.AddFunction(IN_VARIANT(VARCHAR, BOOLEAN));
+	set.AddFunction(IN_VARIANT(BOOLEAN, VARCHAR));
+
+	set.AddFunction(IN_VARIANT(TIMESTAMP, TIMESTAMP));
+	set.AddFunction(IN_VARIANT(TIMESTAMP, VARCHAR));
+	set.AddFunction(IN_VARIANT(VARCHAR, TIMESTAMP));
```
Comment (krlmlr) [inline src/relop.cpp:492]: These two variants should throw a binding error, this means that duckplyr will forward this to R.

File: src/relop.cpp [hunk 3]
```diff
@@ -264,5 +264,241 @@ ScalarFunctionSet base_r_gte() {
...
+
+	set.AddFunction(IN_VARIANT(DATE, DATE));
+	set.AddFunction(IN_VARIANT(DATE, VARCHAR));
+	set.AddFunction(IN_VARIANT(VARCHAR, DATE));
```
Comment (krlmlr) [inline src/relop.cpp:496]: Same here.

File: src/relop.cpp [hunk 4]
```diff
@@ -264,5 +264,241 @@ ScalarFunctionSet base_r_gte() {
...
+	set.AddFunction(IN_VARIANT(DOUBLE, INTEGER));
+	set.AddFunction(IN_VARIANT(INTEGER, DOUBLE));
+	set.AddFunction(IN_VARIANT(DOUBLE, BOOLEAN));
+	set.AddFunction(IN_VARIANT(BOOLEAN, DOUBLE));
+
+	set.AddFunction(IN_VARIANT(VARCHAR, VARCHAR));
+	// set.AddFunction(IN_VARIANT(VARCHAR, INTEGER));
+	// set.AddFunction(IN_VARIANT(INTEGER, VARCHAR));
+	// set.AddFunction(IN_VARIANT(VARCHAR, DOUBLE));
+	// set.AddFunction(IN_VARIANT(DOUBLE, VARCHAR));
```
Comment (krlmlr) [inline src/relop.cpp:486]:
```suggestion
```

File: src/relop.cpp [hunk 5]
```diff
@@ -264,5 +264,241 @@ ScalarFunctionSet base_r_gte() {
...
+	set.AddFunction(IN_VARIANT(VARCHAR, BOOLEAN));
+	set.AddFunction(IN_VARIANT(BOOLEAN, VARCHAR));
```
Comment (krlmlr) [inline src/relop.cpp:488]: Same here.

File: src/relop.cpp [hunk 6]
```diff
@@ -264,5 +264,241 @@ ScalarFunctionSet base_r_gte() {
...
+
+} // namespace
+
+ScalarFunctionSet base_r_in() {
+	ScalarFunctionSet set("r_base::%in%");
+
+	set.AddFunction(IN_VARIANT(DOUBLE, DOUBLE));
+	set.AddFunction(IN_VARIANT(BOOLEAN, BOOLEAN));
+	set.AddFunction(IN_VARIANT(BOOLEAN, INTEGER));
+	set.AddFunction(IN_VARIANT(INTEGER, BOOLEAN));
```
Comment (krlmlr) [inline src/relop.cpp:474]: In general, same-type operations are good, double vs. integer variants are good, for all other combinations we probably want to forward to R.

This is a conclusion I have arrived to only recently, sorry about the extra work.

File: src/relop.cpp [hunk 7]
```diff
@@ -264,5 +264,241 @@ ScalarFunctionSet base_r_gte() {
...
+	set.AddFunction(IN_VARIANT(INTEGER, INTEGER));
+
+	set.AddFunction(IN_VARIANT(DOUBLE, INTEGER));
+	set.AddFunction(IN_VARIANT(INTEGER, DOUBLE));
+	set.AddFunction(IN_VARIANT(DOUBLE, BOOLEAN));
+	set.AddFunction(IN_VARIANT(BOOLEAN, DOUBLE));
```
Comment (krlmlr) [inline src/relop.cpp:480]: Same.
